### ISSUE TITLE
TFP backend - Transformed data

### ISSUE THREAD

**Issue by adamhaber** [stan-dev/stanc3#425]
For the following Stan program:
```stan
data { 
    int<lower=0,upper=1> X; 
} 
transformed data {
    int X2 = X+1;
}
parameters { 
    real<lower=0,upper=1> p; 
} 
model { 
    X2 ~ bernoulli(p); 
} 
```
stan2tfp generates:
```python
class transformed_params_model(tfd__.Distribution):

  def __init__(self, X):
    self.X = X
     
  
  def log_prob_one_chain(self, params):
    target = 0
    X = self.X
    p = tf__.cast(params[0], tf__.float64)
    target += tf__.reduce_sum(tfd__.Bernoulli(None, p).log_prob(X2))
    return target
     
  def log_prob(self, params):
    return tf__.vectorized_map(self.log_prob_one_chain, params)
    
     
  def parameter_shapes(self, nchains__):
    X = self.X
    return [(nchains__, )]
     
  def parameter_bijectors(self):
    X = self.X
    return [tfb__.Chain([tfb__.Shift(tf__.cast(0, tf__.float64)),
                         tfb__.Scale(tf__.cast(1, tf__.float64) - tf__.cast(0, tf__.float64)),
                         tfb__.Sigmoid()])]
     
  def parameter_names(self):
    return ["p"]
     
model = transformed_params_model
```
in which X2 is not defined.

I assume we want to add `self.X2 = self.X + 1` in `__init__`, and then add `X2 = self.X2` everywhere we have `X = self.X`? If that's the case, I'll be happy to try and fix this (presumably by changing `pp_init`?).

**Comment by nhuurre** [comment]
Sounds about right. In the C++ backend model class constructor executes the transformed data block. That corresponds to `__init__()` in Python.
`pp_init` reads data names from `p.input_vars`. The `transformed data` code is `p.prepare_data`.

**Comment by brianwa84** [comment]
Fine for transformed data, but be careful with transformed parameters that
they end up in the scope of logprob, so that tf.GradientTape will properly
record ops for autodiff.

**Comment by adamhaber** [comment]
Thanks @brianwa84 ! Regarding transformed parameters - defining them within the scope of `log_prob_one_chain` will cause their gradients to be recorded properly, but they won't be saved, right? 

@nhuurre - I've tried to fix this by changing `pp_init`'s:
```ocaml
  let ppbody ppf =
    match p.Program.input_vars with
    | [] -> pf ppf "pass"
    | _ -> (list ~sep:cut pp_save_data) ppf p.Program.input_vars
  in
```
to 
```ocaml
 let ppbody ppf =
    match p.Program.input_vars with
    | [] -> pf ppf "pass"
    | _ -> pf ppf "@[<v>%a@,%a@]" (list ~sep:cut pp_save_data) p.Program.input_vars (list ~sep:cut pp_stmt) p.Program.prepare_data
  in

```
This seems to generate the correct `__init__` function; however,

1. It requires returning unit for `| NRFunApp (CompilerInternal, _, _)` in `pp_stmt` (otherwise the `FnChecks` raise an exception). Will this be problematic?
2. All these units are translated to multiple empty lines in `__init__`; any easy way to avoid printing them? 

Thanks!

**Comment by nhuurre** [comment]
I don't know any easy fix for the empty lines.

As for (1.) I assume it works when the data doesn't have `<lower=0,upper=1>`constraint? I just checked and turns out constraints in `transformed parameters` also fail.
I think the general solution (other than just implementing the check functions) is to remove these in Transform_mir.ml. There `remove_unused_stmts` could have
```ocaml
    | NRFunApp (CompilerInternal, name, _)
      when Internal_fun.to_string FnCheck = name ->
       Stmt.Fixed.Pattern.Skip
```

**Comment by brianwa84** [comment]
> Regarding transformed parameters - defining them within the scope of log_prob_one_chain will cause their gradients to be recorded properly, but they won't be saved, right?

That sounds correct to me. IIRC pymc4 approaches this using a `Deterministic` distribution which captures the transformed value.

**Comment by adamhaber** [comment]
@nhuurre - thanks, this seems to work better. Maybe instead of skipping these bound checks, we can translate them to assert statements in `__init__`?

**Comment by nhuurre** [comment]
Good idea. It works for lower/upper bounds but I don't think there's any way to check ordered vector or correlation matrix.
Changing `pp_stmt` also adds `transformed parameters` asserts to `log_prob_one_chain`.

**Comment by adamhaber** [comment]
Wouldn't `np.all(np.diff(x)>0)` (or something similar) work for ordered vectors? I'll try to add the lower/upper bound asserts after stan-dev/stanc3#421  and stan-dev/stanc3#426 .

> Changing pp_stmt also adds transformed parameters asserts to log_prob_one_chain.

I'm not sure I understand; Do you mean that adding a case for `FnCheck` (or other checks) in `pp_stmt` (instead of skipping them in `remove_unused_stmts`) would generate assert statements in `log_prob_one_chain` (and `fun_def`)? I assume we only really want asserts in `__init__`, so we can change `pp_stmt` to take something like an `is_init` bool and only generate these if it's true...

**Comment by nhuurre** [comment]
> Wouldn't np.all(np.diff(x)>0) (or something similar) work for ordered vectors?

Yeah, that would work.

> Do you mean that adding a case for FnCheck (or other checks) in pp_stmt (instead of skipping them in remove_unused_stmts) would generate assert statements in log_prob_one_chain (and fun_def)?

Yes, `pp_stmt` is used for all statement blocks. (of course this only affects models that have constraints in transformed parameters block)

> we can change pp_stmt to take something like an is_init bool and only generate these if it's true

I'm not a fan of complicating `pp_stmt`. `Transform_mir` could remove the constraint checks from `p.log_prob`. I think something like this in `trans_prog` should do it
```ocaml
  {p with log_prob= remove_checks p.log_prob}
  |> Program.map translate_funapps map_stmt
```

**Comment by brianwa84** [comment]
There is an Ordered bijector you could use to ensure a vector is ordered.
For correlation, you could use a CorrelationCholesky bijector. Both will
map from unconstrained reals to the constrained space.

**Comment by adamhaber** [comment]
> Yeah, that would work.

Should be doable to construct similar assertions for other constraints (for example, `np.all(np.linalg.eigvals(x)>-eps)` for covariance matrices, etc).

> Transform_mir could remove the constraint checks from p.log_prob. I think something like this in trans_prog should do it

And similarly for p.functions_block, right?

> There is an Ordered bijector you could use to ensure a vector is ordered.
For correlation, you could use a CorrelationCholesky bijector. Both will
map from unconstrained reals to the constrained space.

Hmm, do you mean using the bijectors _assertions/validate_args functionality to make sure the inputs are valid?

**Comment by nhuurre** [comment]
> And similarly for p.functions_block, right?

Local variables cannot have constraints so there's nothing to remove in `functions_block`.

**Comment by bob-carpenter** [comment]
> On Jan 14, 2020, at 6:36 AM, Brian Patton <notifications@github.com> wrote:
> 
> There is an Ordered bijector you could use to ensure a vector is ordered.
> For correlation, you could use a CorrelationCholesky bijector. Both will
> map from unconstrained reals to the constrained space.

That's how Stan's always done this for parameters (though not all of our transforms are literally bijections---unit_vector using overparameterization plus penalty to get the right uniform distribution in the constrained space). 

In other blocks, constraints are just validated at the end of the block.

**Comment by adamhaber** [comment]
This was closed due to stan-dev/stanc3#426 being merged, right? If so - should I open a new issue to address the bound checks discussed above?

**Comment by nhuurre** [comment]
Right. I didn't check if the PR implements all the details discussed, only that the central issue had been addressed. A new issue for bounds-checks-to-asserts is most welcome.
